ssl/ech.c

```diff
@@ -0,0 +1,5905 @@
+# else
+    pemmod = pemstat.st_mtim.tv_sec;
+# endif
+
+    /*
+     * search list of already loaded keys to see if we have
+     * a macthing one already
+     */
+    pemlen = strlen(pemfname);
+    for (ind = 0; ind != ctx->ext.nechs; ind++) {
+        if (ctx->ext.ech[ind].pemfname == NULL)
+            return OSSL_ECH_KEYPAIR_ERROR;
+        if (pemlen == strlen(ctx->ext.ech[ind].pemfname)
+            && !strncmp(ctx->ext.ech[ind].pemfname, pemfname, pemlen)) {
+            /* matching files! */
+            if (ctx->ext.ech[ind].loadtime < pemmod) {
+                /* aha! load it up so */
+                *index = ind;
+                return OSSL_ECH_KEYPAIR_MODIFIED;
+            } else {
+                /* tell caller no need to bother */
+                *index = -1; /* just in case:-> */
+                return OSSL_ECH_KEYPAIR_UNMODIFIED;
+            }
+        }
+    }
+    *index = -1; /* just in case:-> */
+    return OSSL_ECH_KEYPAIR_NEW;
+}
+
+/*
+ * @brief decode the DNS name in a binary RRData
+ * @param buf points to the buffer (in/out)
+ * @param remaining points to the remaining buffer length (in/out)
+ * @param dnsname returns the string form name on success
+ * @return is 1 for success, error otherwise
+ *
+ * The encoding here is defined in
+ * https://tools.ietf.org/html/rfc1035#section-3.1
+ *
+ * The input buffer pointer will be modified so it points to
+ * just after the end of the DNS name encoding on output. (And
+ * that's why it's an "unsigned char **" :-)
+ */
+static int local_decode_rdata_name(unsigned char **buf, size_t *remaining,
+                                   char **dnsname)
+{
+    unsigned char *cp = NULL;
+    int rem = 0;
+    char *thename = NULL, *tp = NULL;
+    unsigned char clen = 0; /* chunk len */
+
+    if (buf == NULL || remaining == NULL || dnsname == NULL)
+        return 0;
+    rem = (int)*remaining;
+    thename = OPENSSL_malloc(TLSEXT_MAXLEN_host_name);
+    if (thename == NULL)
+        return 0;
+    cp = *buf;
+    tp = thename;
+    clen = *cp++;
+    rem -= 1;
+    if (clen == 0) {
+        /* special case - return "." as name */
+        thename[0] = '.';
+        thename[1] = 0x00;
+    }
+    while (clen != 0 && rem > 0) {
+        if (clen > rem) {
+            OPENSSL_free(thename);
+            return 0;
+        }
+        if (((tp - thename) + clen + 1) > TLSEXT_MAXLEN_host_name) {
+            OPENSSL_free(thename);
+            return 0;
+        }
+        memcpy(tp, cp, clen);
+        tp += clen;
+        *tp++ = '.';
+        cp += clen;
+        rem -= (clen + 1);
+        if (rem <= 0) {
+            OPENSSL_free(thename);
+            return 0;
+        }
+        clen = *cp++;
+    }
+    *buf = cp;
+    if (rem <= 0) {
+        OPENSSL_free(thename);
+        return 0;
+    }
+    *remaining = rem;
+    *dnsname = thename;
+    return 1;
+}
+
+/*
+ * @brief Try figure out ECHConfig encodng by looking for telltales
+ * @param eklen is the length of rrval
```

Comment:
Can you explain the "|thing|" notation to me? I've seen it in comments but am not clear when it's supposed to be used and, since I guess this'd be a thing that'd need doing a bunch of times, I'd prefer understand it better before starting.